jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java

```diff
@@ -101,6 +107,123 @@ public interface HttpChannel extends Invocable
+     * <p>Listener instances that are set as a bean on the {@link Connector} are
+     * efficiently added to {@link HttpChannel}.
+     */
+    public interface Listener extends EventListener
+    {
+        /**
+         * Invoked just after the HTTP request line and headers have been parsed.
+         *
+         * <p>
+         * This the state of the request from the network, and does not include
+         * any request customizations (eg: forwarding, secure, etc)
+         * </p>
+         *
+         * <p>
+         * Implementations must not block!
+         * </p>
+         *
+         * @param request the request object
+         */
+        default void onRequestBegin(Request request)
+        {
+        }
+
+        /**
+         * Invoked just before calling the server handler tree.
+         *
+         * <p>
+         * This is the final state of the request before the handlers are called.
+         * This includes any request customization.
+         * </p>
+         *
+         * @param request the request object
+         */
+        default void onBeforeHandling(Request request)
+        {
+        }
+
+        /**
+         * Invoked after application handling
+         *
+         * @param request the request object
+         * @param handled if the server handlers handled the request
+         * @param failure the exception thrown by the application
+         * @see Handler#handle(Request, Response, Callback)
+         */
+        default void onAfterHandling(Request request, boolean handled, Throwable failure)
+        {
+        }
+
+        /**
+         * Invoked every time a request content chunk has been parsed, just before
+         * making it available to the application.
+         *
+         * <p>
+         *     TODO: make notes about Trailers / EOF / Errors
+         * </p>
+         *
+         * @param request the request object
+         * @param chunk a request content chunk TODO: make bytebuffer a slice?
+         */
+        default void onRequestRead(Request request, Content.Chunk chunk)
+        {
+        }
+
+        /**
+         * Invoked just before the response line is written to the network.
+         *
+         * @param request the request object
+         * @param status the response status
+         * @param response the response object
+         */
+        default void onResponseCommitted(Request request, int status, HttpFields response)
+        {
+        }
+
+        /**
+         * Invoked after a response content chunk has been written to the network.
+         *
+         * @param request the request object
+         * @param last indicating last write
+         * @param content a {@link ByteBuffer#slice() slice} of the response content chunk
+         * @param failure if there was a failure to write the given content
+         */
+        default void onResponseWrite(Request request, boolean last, ByteBuffer content, Throwable failure)
```

Comment:
I don't like the expense of calling this one. I'll do a branch suggesting an alternative... Stand by...

Reply:
@joakime see #9872 for a cheaper implementation for write events (split into two). Also I improved some of the comments